File: apps/api/src/media/signed-url.ts
import { createHmac, timingSafeEqual } from "node:crypto";

const DEFAULT_TTL_SEC = 60 * 60;

function signingSecret(explicit?: string): string {
  const secret = explicit ?? process.env.MEDIA_SIGNING_SECRET ?? process.env.JWT_SECRET;
  if (!secret) throw new Error("MEDIA_SIGNING_SECRET or JWT_SECRET must be set");
  return secret;
}

function hmacHex(secret: string, assetId: string, exp: number): string {
  return createHmac("sha256", secret).update(`${assetId}:${exp}`).digest("hex");
}

export function signMediaUrl(input: {
  assetId: string;
  baseUrl: string;
  secret?: string;
  nowSec?: number;
  ttlSec?: number;
}): string {
  const nowSec = input.nowSec ?? Math.floor(Date.now() / 1000);
  const ttlSec = input.ttlSec ?? DEFAULT_TTL_SEC;
  const exp = nowSec + ttlSec;
  const sig = hmacHex(signingSecret(input.secret), input.assetId, exp);
  const base = input.baseUrl.replace(/\/$/, "");
  return `${base}/media/${input.assetId}?exp=${exp}&sig=${sig}`;
}

export function publicApiBaseUrl(): string {
  const baseUrl = process.env.API_PUBLIC_URL;
  if (!baseUrl) throw new Error("API_PUBLIC_URL must be set");
  return baseUrl.replace(/\/$/, "");
}

export function signedPlaybackForAsset(assetId: string): string {
  return signMediaUrl({ assetId, baseUrl: publicApiBaseUrl() });
}

export function signedHlsForAsset(assetId: string, file = "index.m3u8"): string {
  return signHlsUrl({ assetId, baseUrl: publicApiBaseUrl(), file });
}

export function signHlsUrl(input: {
  assetId: string;
  baseUrl: string;
  file?: string;
  secret?: string;
  nowSec?: number;
  ttlSec?: number;
}): string {
  const unsigned = signMediaUrl({ ...input, ttlSec: input.ttlSec, nowSec: input.nowSec, secret: input.secret });
  const u = new URL(unsigned);
  const file = input.file ?? "index.m3u8";
  u.pathname = `/media/${input.assetId}/hls/${file}`;
  return u.toString();
}

export function verifyMediaSig(input: {
  assetId: string;
  exp: number;
  sig: string;
  secret?: string;
  nowSec?: number;
}): boolean {
  if (!Number.isFinite(input.exp) || input.sig.length !== 64) return false;
  const nowSec = input.nowSec ?? Math.floor(Date.now() / 1000);
  if (input.exp < nowSec) return false;
  const expected = hmacHex(signingSecret(input.secret), input.assetId, input.exp);
  try {
    return timingSafeEqual(Buffer.from(expected, "hex"), Buffer.from(input.sig, "hex"));
  } catch {
    return false;
  }
}

File: apps/api-python/src/jplearn_api/adapters/security/signed_url.py
from __future__ import annotations

import hashlib
import hmac
from urllib.parse import urlparse, urlunparse

DEFAULT_TTL_SEC = 60 * 60


def signing_secret(explicit: str | None, media_signing_secret: str | None, jwt_secret: str) -> str:
    secret = explicit or media_signing_secret or jwt_secret
    if not secret:
        raise RuntimeError("MEDIA_SIGNING_SECRET or JWT_SECRET must be set")
    return secret


def hmac_hex(secret: str, asset_id: str, exp: int) -> str:
    return hmac.new(secret.encode("utf-8"), f"{asset_id}:{exp}".encode("utf-8"), hashlib.sha256).hexdigest()


def sign_media_url(
    *,
    asset_id: str,
    base_url: str,
    secret: str,
    now_sec: int,
    ttl_sec: int = DEFAULT_TTL_SEC,
) -> str:
    exp = now_sec + ttl_sec
    sig = hmac_hex(secret, asset_id, exp)
    base = base_url.rstrip("/")
    return f"{base}/media/{asset_id}?exp={exp}&sig={sig}"


def sign_hls_url(
    *,
    asset_id: str,
    base_url: str,
    secret: str,
    now_sec: int,
    file: str = "index.m3u8",
    ttl_sec: int = DEFAULT_TTL_SEC,
) -> str:
    unsigned = sign_media_url(
        asset_id=asset_id,
        base_url=base_url,
        secret=secret,
        now_sec=now_sec,
        ttl_sec=ttl_sec,
    )
    parsed = urlparse(unsigned)
    return urlunparse(parsed._replace(path=f"/media/{asset_id}/hls/{file}"))


def verify_media_sig(
    *,
    asset_id: str,
    exp: int,
    sig: str,
    secret: str,
    now_sec: int,
) -> bool:
    if not isinstance(exp, int) and not (isinstance(exp, float) and exp.is_integer()):
        return False
    exp_int = int(exp)
    if len(sig) != 64:
        return False
    if exp_int < now_sec:
        return False
    expected = hmac_hex(secret, asset_id, exp_int)
    try:
        return hmac.compare_digest(bytes.fromhex(expected), bytes.fromhex(sig))
    except ValueError:
        return False
